tests/rptest/tests/consumer_group_recovery_tool_test.py

```diff
@@ -31,8 +31,9 @@ def __init__(self, test_ctx, *args, **kwargs):
             test_ctx,
             num_brokers=3,
             *args,
-            # disable leader balancer to make sure that group will not be realoaded because of leadership changes
+            # disable leader balancer to make sure that group will not be reloaded or removed because of leadership changes
```

Comment:
Would a group be removed because of leadership change? It should just migrate to new group leader

Reply:
Yes. I explained in detail in https://github.com/redpanda-data/redpanda/issues/10363#issuecomment-1642598711

But the TLDR is that, as a result of a leadership change, the kafka group manager could invoke `gc_partition_state()` ([here](https://github.com/redpanda-data/redpanda/blob/3e0311cfa0c842cac16b1d6bb0682265a33f26e8/src/v/kafka/server/group_manager.cc#L610) and that method will remove the consumer group ([here](https://github.com/redpanda-data/redpanda/blob/3e0311cfa0c842cac16b1d6bb0682265a33f26e8/src/v/kafka/server/group_manager.cc#L558)).

Reply:
But doesn't that just remove it from that group manager, and eventually it will be added to another group manager, the node that ends up being the new leader for the consumer groups topic/partition that had been moved

Reply:
> hmm. I don't know what to say here since the logs indicate that a leader change caused group removal and the group was **not** recreated/added to another manager.
> 
> Let me see if I can repro by forcing a leader change

I was able to repro the failure locally by forcing a leadership change on `__consumer_offsets/11`. I posted the repro at https://github.com/redpanda-data/redpanda/pull/12366#issuecomment-1644579221

CC @graphcareful

Reply:
So #2217 cleans up old groups that should have been removed because leadership had changed, moved a group to a new leader, then was never removed on the previous leader.

I think if you find some evidence that shows that eventually a new leader was chosen for `__consumer_groups/11` then you can safely close this and implement this solution, but if there is no new leader for this group then there should be a bug right?

Reply:
I think @graphcareful raises a good point.  We should ensure that all groups recover and get a new leader when leadership balancing is enabled.

Reply:
> So #2217 cleans up old groups that should have been removed because leadership had changed, moved a group to a new leader, then was never removed on the previous leader.
> 
> I think if you find some evidence that shows that eventually a new leader was chosen for `__consumer_groups/11` then you can safely close this and implement this solution, but if there is no new leader for this group then there should be a bug right?

Excellent point @graphcareful . I checked the logs of one of the failed tests in #12366 , here are the results from **test run 3** in https://buildkite.com/redpanda/redpanda/builds/33686#0189753e-d0ff-492d-bf12-68f5927f72d1 

## Results

The original leader of `__consumer_offsets/11` is node 3 (docker-rp-11) and the new leader will be node 2 (docker-rp-10)

1. Node 3, the original leader, is removing the consumer group from memory as apart of leadership change
```
INFO  2023-07-20 21:43:16,523 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] consensus.cc:203 - [leadership_transfer] Stepping down as leader in term 1, dirty offset 17
DEBUG 2023-07-20 21:43:16,524 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] consensus.cc:2853 - triggering leadership notification with term: 1, new leader: {nullopt}
TRACE 2023-07-20 21:43:16,525 [shard 0] cluster - metadata_dissemination_service.cc:185 - updating {kafka/__consumer_offsets/11} leadership locally
TRACE 2023-07-20 21:43:16,525 [shard 0] cluster - partition_leaders_table.cc:202 - updated partition: {kafka/__consumer_offsets/11} leader: {term: 1, current leader: {nullopt}, previous leader: {3}, revision: 37}
INFO  2023-07-20 21:43:16,525 [shard 0] kafka - group_manager.cc:559 - Removing groups of {kafka/__consumer_offsets/11}
TRACE 2023-07-20 21:43:16,526 [shard 1] cluster - partition_leaders_table.cc:202 - updated partition: {kafka/__consumer_offsets/11} leader: {term: 1, current leader: {nullopt}, previous leader: {3}, revision: 37}
INFO  2023-07-20 21:43:16,526 [shard 0] kafka - group_manager.cc:573 - Removed group id={kgo-verifier-1689889371-72-0} state=Empty gen=0 proto_type={nullopt} proto={nullopt} leader={nullopt} empty=true ntp={kafka/__consumer_offsets/11} num_members_joining=0 new_member_added=false join_timer={nullopt} pending members [] full members []
```


2. Node 2 is sending vote requests to become the new leader
```
INFO  2023-07-20 21:43:16,533 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:52 - Sending vote request to {id: {3}, revision: {37}} with timeout 1500
INFO  2023-07-20 21:43:16,534 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:52 - Sending vote request to {id: {1}, revision: {37}} with timeout 1500
```

3. Other nodes, such as Node 3 here, recieved the vote request from node 2
```
TRACE 2023-07-20 21:43:16,538 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] consensus.cc:1629 - Received vote request: {node_id: {id: {2}, revision: {37}}, target_node_id{id: {3}, revision: {37}}, group: {38}, term:{2}, prev_log_index:{17}, prev_log_term: {1}, leadership_xfer: 1}
INFO  2023-07-20 21:43:16,538 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] consensus.cc:1718 - Received vote request with larger term from node {id: {2}, revision: {37}}, received 2, current 1
```

4. Node 2 gets response from others and becomes the new leader
```
INFO  2023-07-20 21:43:16,553 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:77 - vote reply from {id: {1}, revision: {37}} - {term:{2}, target_node_id{id: {2}, revision: {37}}, vote_granted: 1, log_ok:1}
TRACE 2023-07-20 21:43:16,554 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:256 - vote acks in term 2 from: {{id: {2}, revision: {37}}, {id: {1}, revision: {37}}}
INFO  2023-07-20 21:43:16,554 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:265 - becoming the leader term:2
DEBUG 2023-07-20 21:43:16,554 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] consensus.cc:2375 - Replicating group configuration {current: {voters: {{id: {3}, revision: {37}}, {id: {2}, revision: {37}}, {id: {1}, revision: {37}}}, learners: {}}, old:{nullopt}, revision: 37, update: {nullopt}, version: 5}}
TRACE 2023-07-20 21:43:16,556 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] replicate_entries_stm.cc:193 - Self append entries - {raft_group:{38}, commit_index:{17}, term:{2}, prev_log_index:{17}, prev_log_term:{1}, last_visible_index:{17}}
INFO  2023-07-20 21:43:16,559 [shard 0] raft - [group_id:38, {kafka/__consumer_offsets/11}] vote_stm.cc:77 - vote reply from {id: {3}, revision: {37}} - {term:{2}, target_node_id{id: {2}, revision: {37}}, vote_granted: 1, log_ok:1}
```

5. The DT test issues list groups request before node 2 finishes replicating the configuration. This leads to test failure because, for this brief moment, none of the brokers hold the consumer group in memory (cg recovery is still happening).
```
[DEBUG - 2023-07-20 21:43:16,528 - rpk - _execute - lineno:895]: Executing command: ['/var/lib/buildkite-agent/builds/buildkite-amd64-xfs-builders-i-0ef6bd5f8613f337f-1/redpanda/redpanda/vbuild/redpanda_installs/ci/bin/rpk', 'group', '-X', 'brokers=docker-rp-11:9092,docker-rp-9:9092,docker-rp-10:9092', 'list', '-v']
[DEBUG - 2023-07-20 21:43:16,604 - rpk - _execute - lineno:911]: 
BROKER  GROUP
```

6. RP initiates shutdown because DT stopped because the test recieved empty ListGroups response from step 5
```
INFO  2023-07-20 21:43:16,948 [shard 0] main - application.cc:379 - Stopping...
INFO  2023-07-20 21:43:16,948 [shard 0] kafka - server.cc:278 - kafka rpc protocol - Stopping 3 listeners
DEBUG 2023-07-20 21:43:16,949 [shard 0] kafka - server.cc:282 - kafka rpc protocol - Service probes 0x6080000738a0
INFO  2023-07-20 21:43:16,949 [shard 0] kafka - server.cc:287 - kafka rpc protocol - Shutting down 0 connections
DEBUG 2023-07-20 21:43:16,949 [shard 1] rpc - transport.cc:385 - could not parse header from server: {host: docker-rp-9, port: 33145}
INFO  2023-07-20 21:43:16,949 [shard 1] kafka - server.cc:278 - kafka rpc protocol - Stopping 3 listeners
DEBUG 2023-07-20 21:43:16,950 [shard 0] rpc - transport.cc:385 - could not parse header from server: {host: docker-rp-9, port: 33145}
DEBUG 2023-07-20 21:43:16,950 [shard 1] kafka - server.cc:282 - kafka rpc protocol - Service probes 0x608000075320
INFO  2023-07-20 21:43:16,950 [shard 1] kafka - server.cc:287 - kafka rpc protocol - Shutting down 0 connections
```

Therefore, I believe a different broker was becoming the new leader for `__consumer_offsets/11` but the test issued a ListGroups request before recovery finished. So, re-writing this test to disable leadership tranfers will solve the CI failure. However, IMO, we should still test that consumer group recovery works in the presence of a leadership change on a `__consumer_offsets` NTP. So, I will also modify the python script to test that scenario as a separate DT test.

Reply:
Real nice write up and discovery ! +1 I agree with your next steps as well

Reply:
I think in this case handler should return coordinator_load_in_progress this way client would now that it must retry

Reply:
Spoke with @mmaslankaprv  offline.

The concern is at step (5).

> This leads to test failure because, for this brief moment, none of the brokers hold the consumer group in memory (cg recovery is still happening).

If there is no leader for a consumer offsets NTP in this brief moment, then `coordinator_load_in_progress` should be returned. However, that is not happening.
But does upstream Kafka also respond with an empty list_groups in the same scenario?